File: rxbilling-google-play-library-v3/src/main/java/com/vanniktech/rxbilling/google/play/library/PlayBillingInventorySubscription.java
package com.vanniktech.rxbilling.google.play.library;

import androidx.annotation.NonNull;
import com.android.billingclient.api.SkuDetails;
import com.google.auto.value.AutoValue;
import com.google.auto.value.extension.memoized.Memoized;
import com.vanniktech.rxbilling.InventorySubscription;
import java.math.BigDecimal;
import java.util.Currency;

import static com.vanniktech.rxbilling.google.play.library.Utils.asBigDecimal;

@SuppressWarnings("PMD.UseObjectForClearerAPI") @AutoValue abstract class PlayBillingInventorySubscription implements InventorySubscription {
  @NonNull public static InventorySubscription create(@NonNull final SkuDetails skuDetails) {
    return new AutoValue_PlayBillingInventorySubscription(skuDetails.getSku(), skuDetails.getType(), skuDetails.getPrice(), skuDetails.getPriceAmountMicros(), skuDetails.getPriceCurrencyCode(),
        skuDetails.getTitle(), skuDetails.getDescription(), skuDetails);
  }

  @Override @NonNull @Memoized public Currency priceCurrency() {
    return Currency.getInstance(priceCurrencyCode());
  }

  @Override @NonNull @Memoized public BigDecimal priceAsBigDecimal() {
    return asBigDecimal(priceAmountMicros());
  }

  @NonNull abstract SkuDetails skuDetails();
}

File: rxbilling-google-play-library-v3/src/main/java/com/vanniktech/rxbilling/google/play/library/PlayBillingInventorySubscription.kt
@file:JvmName("Utils")

package com.vanniktech.rxbilling.google.play.library

import com.android.billingclient.api.SkuDetails
import com.vanniktech.rxbilling.InventorySubscription
import java.math.BigDecimal
import java.util.Currency

data class PlayBillingInventorySubscription(
  val skuDetails: SkuDetails,
) : InventorySubscription {
  override val sku: String = skuDetails.sku
  override val type: String = skuDetails.type
  override val price: String = skuDetails.price
  override val priceAmountMicros: Long = skuDetails.priceAmountMicros
  override val priceCurrencyCode: String = skuDetails.priceCurrencyCode
  override val title: String = skuDetails.title
  override val description: String = skuDetails.description

  override fun priceCurrency(): Currency = Currency.getInstance(priceCurrencyCode)
  override fun priceAsBigDecimal(): BigDecimal = priceAmountMicros.microsAsBigDecimal()
}
